File: src/mcp/tools/definitions/timeline_create.ts
/**
 * timeline_createツール定義
 * タイムライン要素を作成するMCPツール
 */

import type {
  McpToolDefinition,
  ToolExecutionContext,
} from "../tool_registry.ts";
import { executeCliCommand } from "../cli_adapter.ts";
import { ElementTimelineCommand } from "../../../cli/modules/element/timeline.ts";

export const timelineCreateTool: McpToolDefinition = {
  name: "timeline_create",
  description:
    "タイムライン（時系列）要素を作成します。物語の時系列イベントを管理するための構造を定義します。",
  inputSchema: {
    type: "object",
    properties: {
      name: {
        type: "string",
        description: "タイムライン名",
      },
      scope: {
        type: "string",
        enum: ["story", "world", "character", "arc"],
        description:
          "タイムラインのスコープ: story（物語全体）, world（世界史）, character（キャラクター固有）, arc（章・アーク単位）",
      },
      summary: {
        type: "string",
        description: "タイムラインの概要",
      },
      id: {
        type: "string",
        description: "タイムラインID（省略時はnameから自動生成）",
      },
      parentTimeline: {
        type: "string",
        description: "親タイムラインのID（サブタイムラインの場合）",
      },
      relatedCharacter: {
        type: "string",
        description: "関連キャラクターID（キャラクター固有タイムラインの場合）",
      },
      displayNames: {
        type: "array",
        items: { type: "string" },
        description: "表示名のバリエーション",
      },
    },
    required: ["name", "scope"],
  },
  execute: async (
    args: Record<string, unknown>,
    context?: ToolExecutionContext,
  ) => {
    const name = args.name as string | undefined;
    const scope = args.scope as string | undefined;

    if (!name || typeof name !== "string" || name.trim() === "") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'name' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (!scope || typeof scope !== "string") {
      return {
        content: [
          {
            type: "text" as const,
            text:
              "Error: 'scope' parameter is required (story|world|character|arc).",
          },
        ],
        isError: true,
      };
    }

    const validScopes = ["story", "world", "character", "arc"];
    if (!validScopes.includes(scope)) {
      return {
        content: [
          {
            type: "text" as const,
            text: `Error: Invalid scope '${scope}'. Must be one of: ${
              validScopes.join(", ")
            }.`,
          },
        ],
        isError: true,
      };
    }

    // CLIコマンド用の引数を構築
    const commandArgs: Record<string, unknown> = {
      name,
      scope,
    };

    if (typeof args.summary === "string") {
      commandArgs.summary = args.summary;
    }

    if (typeof args.id === "string") {
      commandArgs.id = args.id;
    }

    if (typeof args.parentTimeline === "string") {
      commandArgs["parent-timeline"] = args.parentTimeline;
    }

    if (typeof args.relatedCharacter === "string") {
      commandArgs["related-character"] = args.relatedCharacter;
    }

    if (Array.isArray(args.displayNames)) {
      const displayNames = (args.displayNames as unknown[])
        .filter((n) => typeof n === "string")
        .map((n) => (n as string).trim())
        .filter((n) => n.length > 0);
      if (displayNames.length > 0) {
        commandArgs.displayNames = displayNames.join(",");
      }
    } else if (typeof args.displayNames === "string") {
      commandArgs.displayNames = args.displayNames;
    }

    const handler = new ElementTimelineCommand();
    return await executeCliCommand(handler, commandArgs, context?.projectRoot);
  },
};

File: internal/mcp/tools/timeline_create.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/takets/street-storyteller/internal/domain"
	"github.com/takets/street-storyteller/internal/mcp/protocol"
)

// TimelineCreateTool validates inputs and returns the generated Timeline stub.
// Why: persistence is out of scope for this tool — the caller (CLI/MCP client)
// owns the write path; the tool only generates and validates the structure.
type TimelineCreateTool struct{}

type timelineCreateArgs struct {
	Name    string `json:"name"`
	Scope   string `json:"scope"`
	Summary string `json:"summary"`
	ID      string `json:"id"`
}

var validScopes = map[domain.TimelineScope]struct{}{
	domain.TimelineScopeStory:     {},
	domain.TimelineScopeWorld:     {},
	domain.TimelineScopeCharacter: {},
	domain.TimelineScopeArc:       {},
}

// timelineSlugify converts a human-readable name into a lowercase underscore id.
var timelineSlugRe = regexp.MustCompile(`[^a-z0-9]+`)

func timelineSlugify(name string) string {
	s := strings.ToLower(name)
	s = timelineSlugRe.ReplaceAllString(s, "_")
	s = strings.Trim(s, "_")
	return s
}

// Definition advertises the timeline_create schema.
func (TimelineCreateTool) Definition() protocol.Tool {
	return protocol.Tool{
		Name:        "timeline_create",
		Description: "Create a new timeline and return its structure as JSON",
		InputSchema: json.RawMessage(`{
			"type":"object",
			"required":["name","scope","summary"],
			"properties":{
				"name":{"type":"string","description":"Human-readable timeline name"},
				"scope":{"type":"string","enum":["story","world","character","arc"],"description":"Timeline scope"},
				"summary":{"type":"string","description":"Brief summary of the timeline"},
				"id":{"type":"string","description":"Optional ID; derived from name if omitted"}
			}
		}`),
	}
}

// Handle validates args and returns the generated Timeline stub as text.
func (TimelineCreateTool) Handle(_ context.Context, args json.RawMessage, _ ExecutionContext) (*protocol.CallToolResult, error) {
	var a timelineCreateArgs
	if len(args) > 0 {
		_ = json.Unmarshal(args, &a)
	}

	if a.Name == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "name is required"}},
			IsError: true,
		}, nil
	}
	if a.Scope == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "scope is required"}},
			IsError: true,
		}, nil
	}
	if a.Summary == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "summary is required"}},
			IsError: true,
		}, nil
	}

	scope := domain.TimelineScope(a.Scope)
	if _, ok := validScopes[scope]; !ok {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: fmt.Sprintf("invalid scope %q: must be one of story, world, character, arc", a.Scope)}},
			IsError: true,
		}, nil
	}

	id := a.ID
	if id == "" {
		id = timelineSlugify(a.Name)
	}

	tl := domain.Timeline{
		ID:      id,
		Name:    a.Name,
		Scope:   scope,
		Summary: a.Summary,
		Events:  []domain.TimelineEvent{},
	}

	b, _ := json.Marshal(tl)
	return &protocol.CallToolResult{
		Content: []protocol.ContentBlock{{
			Type: "text",
			Text: fmt.Sprintf("timeline created: %s\n%s", id, string(b)),
		}},
	}, nil
}
